sw/device/lib/dif/dif_padctrl.h

```diff
@@ -0,0 +1,196 @@
+// we just autogen the same number twice?
+#define DIF_PADCTRL_DIO_NUM 32
+#define DIF_PADCTRL_MIO_NUM 32
+#define DIF_PADCTRL_ATTR_WIDTH 8
+#define DIF_PADCTRL_ADDITIONAL_ATTR_WIDTH 2
+
+/**
+ * Pad attributes
+ *
+ * Some attributes have defined meanings, others are implementation defined
+ * (|kDifPadCtrlAttrAdditionalFirst| and onwards). not all pads will support all
+ * attributes.
+ */
+typedef enum dif_padctrl_attr {
+  kDifPadctrlAttrIOInvert = 0,
+  kDifPadctrlAttrOpenDrain,
+  kDifPadctrlAttrPullDown,
+  kDifPadctrlAttrPullUp,
+  kDifPadctrlAttrKeeper,
+  kDifPadctrlAttrWeakDrive
+  kDifPadctrlAttrAdditionalFirst = 0,
+  kDifPadctrlAttrAdditionalLast =
+    DIF_PADCTRL_ADDITIONAL_ATTR_WIDTH == 0 ?
+      kDifPadctrlAdditionalAttrFirst :
+      kDifPadctrlAdditionalAttrFirst + DIF_PADCTRL_ADDITIONAL_ATTR_WIDTH - 1,
+} dif_padctrl_attr_t;
+
+/**
+ * DIO pads
+ */
+typedef enum dif_padctrl_dio {
+  kDifPadctrlDioFirst = 0,
+  kDifPadctrlDioLast = DIF_PADCTRL_DIO_NUM - 1,
+} dif_padctrl_dio_t;
+
+/**
+ * MIO Pads
+ */
+typedef enum dif_padctrl_mio {
+  kDifPadctrlMioFirst = 0,
+  kDifPadctrlMioLast = DIF_PADCTRL_MIO_NUM - 1,
+} dif_padctrl_mio_t;
+
+/**
+ * Padctrl instance state.
+ *
+ * Padctrl persistent data that is required by the Padctrl API.
+ */
+typedef struct dif_padctrl {
+  mmio_region_t base_addr; /**< Padctrl base address. */
+  bool write_enabled;      /**< Configuration registers locked/unlocked. */
+} dif_padctrl_t;
+
+/**
+ * Padctrl init routine error codes.
+ */
+typedef enum dif_padctrl_init_error {
+  kDifPadctrlInitErrorNone = 0         /**< Success. */
+  kDifPadctrlInitErrorInvalidArgument, /**< Input parameter is not valid. */
+  kDifPadctrlInitErrorRegistersLocked, /**< Peripheral is locked and cannot be
+                                          reset. */
+} dif_padctrl_init_error_t;
+
+/**
+ * Initialises an instance of Padctrl
+ *
+ * Information that must be retained, and is required to program Padctrl, shall
+ * be stored in @p padctrl.
+ *
+ * @param base_addr Base address of an isntance of the Padctrl IP block
+ * @param padctrl Padctrl state data
+ * @return dif_padctrl_init_error_t.
+ */
+dif_padctrl_init_error_t dif_padctrl_init(mmio_region_t base_addr,
+                                          dif_padctrl_t *padctrl);
+
+/**
+ * Padctrl registers lock routine error codes.
+ */
+typedef enum dif_padctrl_registers_lock_error {
+  kDifPadctrlRegistersLockErrorNone = 0,        /**< Success. */
+  kDifPadctrlRegistersLockErrorInvalidArgument, /**< Input parameter is not
+                                                   valid. */
+} dif_padctrl_registers_lock_error_t;
+
+/**
+ * Locks the padctrl registers.
+ *
+ * After the registers have been locked, they can only be unlocked by the
+ * system reset.
+ *
+ * @param padctrl Padctrl state data.
+ * @return dif_padctrl_registers_lock_error_t.
+ */
+dif_pinmux_registers_lock_error_t dif_pinmux_registers_lock(
+    dif_pinmux_t *pinmux);
```

Comment:
Why does this take *just* a pinmux? Seems odd to me.

Reply:
Indeed, I copied this over from pinmux and neglected to fix the naming.